dotnet/targets/Xamarin.Shared.Sdk.targets

```diff
@@ -1583,7 +1583,7 @@
 			<_DefaultLinkMode>TrimMode</_DefaultLinkMode>
 		</PropertyGroup>
 		<PropertyGroup Condition="'$(TrimMode)' == ''">
-			<_DefaultLinkMode Condition="'$(_UseNativeAot)' != 'true'">Full</_DefaultLinkMode> <!-- Linking is always on for all assemblies when using NativeAOT - this is because we need to modify all assemblies in the linker for them to be compatible with NativeAOT -->
+			<_DefaultLinkMode Condition="'$(_UseNativeAot)' == 'true'">Full</_DefaultLinkMode> <!-- Linking is always on for all assemblies when using NativeAOT - this is because we need to modify all assemblies in the linker for them to be compatible with NativeAOT -->
```

Comment:
I agree that the comment does not match how `_DefaultLinkMode` is set, but I think there is more to this and that the introduced change only works because we explicitly pass: 
https://github.com/xamarin/xamarin-macios/blob/7ed8c08ab0c2080124e171da835506fcfc21c967/dotnet/targets/Xamarin.Shared.Sdk.targets#L1215

Reply:
Without this fix `_DefaultLinkMode` is not set. Then you get to the `_ComputeLinkMode` target here:

https://github.com/xamarin/xamarin-macios/blob/05502d1260bf70d6dc628d2ec5566810dd760f46/msbuild/Xamarin.Shared/Xamarin.Shared.targets#L2263-L2271

On iOS-like platform you end up in the `SdkOnly` code path which translates to `partial` TrimMode iirc and somehow ends up doing mostly the right thing. On macOS, however, you get the `None` mode in the fallback. This result in `copy` TimeMode which does no linking at all and then passes all the assemblies as `--root:<assembly name>` to ILCompiler. That in turns results in compilation of the whole world, and in Debug version of ILC it assert on trying to compile `Array<T>`. That's definitely wrong.

This is all irrespective of `--defaultrooting` which is already files as issue #18482.

Reply:
What I meant is (considering just iOS):
- if `_DefaultLinkMode=Full` -> `_LinkMode` will become `Full`
- if `_LinkMode=Full` -> `TrimMode` will become `full` 
https://github.com/xamarin/xamarin-macios/blob/7ed8c08ab0c2080124e171da835506fcfc21c967/dotnet/targets/Xamarin.Shared.Sdk.targets#L522
- if `TrimMode=full` -> ILC will **not** use `--defaultrooting`
https://github.com/dotnet/runtime/blob/f25bc7b5b859935192e4eddc5b71efe71a60a4de/src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets#L270
- If `--defaultrooting` is not passed to ILC -> ILC will start trimming (which will break the app)

So I just wanted to draw the attention that the above change only works in the current setup, as we are explicitly passing `--defaultrooting` to ILC

Reply:
Additionally, this means that for iOS/tvOS/watchOS (in case of NativeAOT) the default linking mode becomes `Full` which is not how ILLink is configured in the noNativeAOT mode for these platforms: https://github.com/xamarin/xamarin-macios/pull/18530/files#diff-2e24f1b6f44df7580244d00f106fa1a9e2257536fbca9d7bbdc74571549324a9R1591 

This could start breaking the apps with NativeAOT that work normally in noNativeAOT mode, as the change enables ILLink to trim more than what is the current default mode.

Reply:
In line with how PublishAot is treated in dotnet/runtime it would probably make sense to apply the default when PublishAot=true regardless of whether we are actually publishing or not.

Reply:
> So, what do you propose the default to be?

I am not sure. I just think that we should not change ILLink default trimming modes just because NativeAOT compilation later fails. Maybe the right fix would be to "intercept" the build and set the right properties for NativeAOT, after ILLink has finished its work and before ILC compilation starts.

@akoeplinger what do you think?

Reply:
NativeAOT always behaves as a linker though. Running it in non-linking mode is always going to either not work (assert / endless compilation), or - at best - produce completely unusable huge output.

`PublishAot` is an opt-in switch. The general direction on desktop platforms is to run CoreCLR in a mode where it imitates the NativeAOT behavior more closely (eg. no dynamic code, extra analyzer, etc.). The mobile equivalent of that would be to always set the default link mode to the same value when `PublishAot=true`, regardless of `_IsPublishing` and `Configuration`. That way you opt-in with `PublishAot` into the specific behavior and it's honored both for debugging with Mono, and publishing with NativeAOT.

Reply:
> I just think that we should not change ILLink default trimming modes just because NativeAOT compilation later fails.

Also, just in case it's not clear, this is currently changing the default ONLY when `PublishAot=true` AND you are running `dotnet publish`...

> Maybe the right fix would be to "intercept" the build and set the right properties for NativeAOT, after ILLink has finished its work and before ILC compilation starts.

... so the end effect is not really any different than this. It doesn't matter whether ILLink trims it first, or NativeAOT trims it later. In the end you still need the trimming to take place.

Reply:
> Also, just in case it's not clear, this is currently changing the default ONLY when PublishAot=true AND you are running dotnet publish...

Thanks, I did understand this exactly as you explained, but again I just want to emphasise that with this change we will start trimming the user code which was not marked with `IsTrimmable=true` (at least in case of iOS-like platforms) and the reasons for potential regressions that this could bring will not be necessarily related to the fact that we are using NativeAOT as a deployment toolchain-runtime, but rather to the fact that we started trimming something that was never trimmed before.

Reply:
I, however, feel that defaulting to `partial` (iOS-like platforms before this PR) is weird when .NET 7+ default for other platforms is `full` and it's documented that you can change it, and how.

That said, either `partial` or `full` would work to solve the macOS issue.

Reply:
> I, however, feel that defaulting to partial (current default for iOS-like platforms without this PR) is weird when .NET 7+ default for other platforms is full and it's documented that you can change it, and how.

That is something that should be checked with @rolfbjarne

As for the NativeAOT story regarding macOS, I believe something like:
```xml
<_DefaultLinkMode Condition="'$(_UseNativeAot)' == 'true' And '$(_PlatformName)' == 'macOS'">Full</_DefaultLinkMode>
<_DefaultLinkMode Condition="'$(_UseNativeAot)' == 'true' And '$(_PlatformName)' != 'macOS'">SdkOnly</_DefaultLinkMode>
``` 
and
```xml
<IlcArg Condition="''$(_PlatformName)' != 'macOS'" Include="--defaultrooting" />
``` 
should be tested, as in the current set up NativeAOT trimming is disabled, due to: 
1) ILLink removing `IsTrimmable` attributes from assemblies 
2) NativeAOT operating in `--defaultrooting` mode